### ISSUE TITLE
`mtl_buffers_in_flight` membership check causes step time to grow linearly

### ISSUE THREAD

**Issue by Cyberistic** [tinygrad/tinygrad#16595]
On Metal, a workload that emits many command buffers per step (e.g. a graph split into many `MetalGraph` calls, or a per-sample gradient path in my case) grows its per-step latency a lot over a few hundred steps. This affects both `@TinyJit` and eager.

If I call`Device["METAL"].synchronize()` every few iterations, growth stops, which points to the in-flight command-buffer list as the bottleneck.

`tinygrad/runtime/graph/metal.py:53`:

```python
if self.command_buffer is not None and self.command_buffer in self.dev.mtl_buffers_in_flight:
    wait_check(self.command_buffer)
```
O(N) buffers_in_flight membership check, with the list growing with the number of steps since the last sync.

it is appended to in `MetalGraph.__call__` and `MetalProgram.__call__` and `MetalAllocator._transfer`, but only cleared in `MetalDevice.synchronize`

The list should be a set that checks in O(1) membership and auto-removes a command buffer as soon as the GPU finishes it, maybe using `MTLCommandBuffer.addCompletedHandler`?

## Repro

**AI-written** repro given my actual use case as reference. You can also force the behavior by using GRAPH_ONE_KERNEL=1 and JIT_BATCH_SIZE=1

```py

import time
import numpy as np
from tinygrad import Device, Tensor, nn
from tinygrad.nn.state import get_parameters


# ----- minimal DPSGDCNN-like model (copied from the user's script) -----
class DPSGDCNN:
    def __init__(self, num_classes, model_width=16):
        self.conv1 = nn.Conv2d(3, model_width, kernel_size=3, padding=1)
        self.conv2 = nn.Conv2d(model_width, model_width * 2, kernel_size=3, padding=1)
        self.conv3 = nn.Conv2d(model_width * 2, model_width * 4, kernel_size=3, padding=1)
        self.fc1 = nn.Linear(model_width * 4 * 12 * 12, model_width * 8)
        self.fc2 = nn.Linear(model_width * 8, num_classes)

    def __call__(self, x):
        # tinygrad conv2d is NCHW
        if x.shape[-1] == 3 and x.ndim == 4:
            x = x.permute(0, 3, 1, 2).contiguous()
        intermediates = []
        x = self.conv1(x).relu()
        intermediates.append(("conv1", x))
        x = x.max_pool2d(2)
        x = self.conv2(x).relu()
        intermediates.append(("conv2", x))
        x = x.max_pool2d(2)
        x = self.conv3(x).relu()
        intermediates.append(("conv3", x))
        x = x.max_pool2d(2)
        x = x.flatten(1)
        x = self.fc1(x).relu()
        x = self.fc2(x)
        return x, intermediates


# ----- per-example sparse CE loss (copied) -----------------------------
def per_example_sparse_ce(logits, y):
    return logits.gather(1, y.reshape(-1, 1)).squeeze(1) - logits.logsumexp(1)


# ----- ghost step (simplified: per-ex loss + 1st backward + 2nd backward) -----
# This still generates a large number of command buffers per step because
# the two backward passes traverse the graph and produce per-example
# gradients for every parameter and every stashed intermediate.
def ghost_step(model, X, y):
    params = get_parameters(model)
    Tensor.training = True
    for p in params:
        p.grad = None

    logits, intermediates = model(X)
    per_ex_loss = per_example_sparse_ce(logits, y)
    loss_mean = per_ex_loss.mean()

    # 1st backward: gradient of sum-of-per-example-loss w.r.t. every
    # parameter and every stashed conv output. This populates param.grad
    # and produces a gradient tensor for each intermediate.
    targets = list(params) + [t for _, t in intermediates]
    per_ex_loss.sum().gradient(*targets)

    # 2nd backward: same loss, but now the graph is replayed to produce
    # a second set of gradients (in the real ghost path these are
    # weighted by per-example clip factors; we skip that for brevity).
    for p in params:
        p.grad = None
    per_ex_loss.sum().backward()
    return loss_mean


# ----- benchmark --------------------------------------------------------
B = 32
H = W = 100
C = 3
NUM_CLASSES = 10
N_ITERS = 200

np.random.seed(42)
Tensor.manual_seed(42)

X = Tensor(np.random.randn(B, H, W, C).astype(np.float32)).permute(0, 3, 1, 2).contiguous()
y = Tensor(np.random.randint(0, NUM_CLASSES, size=B).astype(np.int32))

model = DPSGDCNN(num_classes=NUM_CLASSES, model_width=16)

# warmup
for _ in range(3):
    loss = ghost_step(model, X, y)
    loss.realize()
    del loss

dev = Device["METAL"]
print(f"after warmup: in_flight={len(dev.mtl_buffers_in_flight)}")
print()
print("iter   step_ms   in_flight")
print("--------------------------------")
times = []
for i in range(N_ITERS):
    t0 = time.perf_counter()
    loss = ghost_step(model, X, y)
    loss.realize()
    t1 = time.perf_counter()
    dt = (t1 - t0) * 1000
    times.append(dt)
    if i % 10 == 0 or i >= N_ITERS - 5:
        print(f"{i:4d}   {dt:7.2f}   {len(dev.mtl_buffers_in_flight):5d}")

t_first = times[1]
t_last = times[-1]
print()
print("=" * 60)
print(f"step time at iter 1   : {t_first:.2f} ms")
print(f"step time at iter {N_ITERS-1} : {t_last:.2f} ms")
print(f"drift                 : +{t_last - t_first:.2f} ms (+{(t_last/t_first - 1)*100:.1f}%)")
```

```
after warmup: in_flight=60

iter   step_ms   in_flight
--------------------------------
   0     23.89      70
  10     25.37     170
  20     26.94     270
  30     27.35     370
  40     28.43     470
  50     29.05     570
  60     30.52     670
  70     31.88     770
  80     33.29     870
  90     34.73     970
 100     35.15    1070
 110     37.14    1170
 120     36.15    1270
 130     38.58    1370
 140     38.55    1470
 150     41.15    1570
 160     42.12    1670
 170     43.51    1770
 180     44.73    1870
 190     44.95    1970
 195     46.93    2020
 196     46.10    2030
 197     47.27    2040
 198     46.70    2050
 199     46.48    2060

============================================================
step time at iter 1   : 26.05 ms
step time at iter 199 : 46.48 ms
drift                 : +20.43 ms (+78.4%)
```

Note that this is a simplified repro which doesn't show the full picture. In my actual use-case it's actually much worse (I can send the repo if necessary):

```
iter   step_ms   in_flight
--------------------------------
   0     83.49     286
  10     85.91    1136
  20    777.50    1986
  30   1342.08    2836
  40    770.65    3686
  50    755.43    4536
  60   1405.86    5386
  70    748.01    6236
  80    777.58    7086
  90   1425.59    7936
 100    802.71    8786
 110   1355.76    9636
 120    701.93   10486
 130    784.28   11336
 140   1326.99   12186
 150   1334.60   13036
 160    709.08   13886
 170   1318.70   14736
 180    792.72   15586
 190    784.00   16436
 195   1312.31   16861
 196   1308.78   16946
 197   1319.57   17031
 198   1318.06   17116
 199   1307.34   17201

============================================================
step time at iter 1   : 78.64 ms
step time at iter 200 : 1307.34 ms
drift                 : +1228.71 ms (+1562.5%)
```
